### ISSUE TITLE
Not correctly selecting VarArgs methods

### ISSUE THREAD

**Issue by opeongo** [beanshell/beanshell#620]
I had a method call on an overridden method blow up with an index out of range exception in the BeanShell code. It seems that the overridden method that was selected had fewer parameters than the arguments that I was passing in.  Surely this should not have happened that the selected method did not match the caller.  After some digging around I came across a two problems.  The code is getting a bit hard to follow, so to keep things as simple as possible I will unpack these problems one by one and propose fixes one by one. 

The first problem I will look at was exposed by a different problem, so I will fix this first before fixing the other problem that will hide this one again.  It looks like the mechanism for selecting the most specific overridden method is not working properly for VarArgs methods.  The rules in[ JLS 15.12.2](https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.12.2) have the order

1. check fixed-arity methods with strict context
2. check fixed-arity methods with loose context (boxing, narrowing)
3. check variable-arity methods with loose context.

As soon as one of these match then the search is over.  However in the current code base the variable arity and fixed arity methods get lumped together at the start of the algorithm and tested simultaneously in steps 1 and 2.

Consider this sample code:
```java
    // for testing VarArgs and most specific method selection
    public String method2(String str) {
        // most specific for method2("Hello")
        return "one";
    }
    public String method2(String str, String str1) {
        // most specific for method2("Hello", "World")
        return "two";
    }
    public String method2(String str, String... str2) {
        // most specific for method2("Hello", "World", "This", "is", "me")
        return "three";
    }
```

The current master BeanShell selects the following:
```java
m.method2("Hello") returns "one"
m.method2("Hello", "World") return "three"
m.method2("Hello", "World", "This", "is", "me") returns "three";
```

I had a look at the code and it seems like there are two separate mechanisms for handling varargs.  The first is started in [Reflect.findMostSpecificInvocable](https://github.com/beanshell/beanshell/blob/cb552a4d08c279680d6463b61e2f034fe12a590b/src/main/java/bsh/Reflect.java#L431).  This one looks like it is on the right track but it does not look like it was fully implemented.  

The second place where varargs are handled in method selection is in [Types.isSignatureVarargsAssignable](https://github.com/beanshell/beanshell/blob/cb552a4d08c279680d6463b61e2f034fe12a590b/src/main/java/bsh/Types.java#L232) but this one is clearly incorrect because it does not consider if a method is marked as `isVarArgs()` and treats all methods with an array in the final parameter position as vararg method candidates (and this is where the unhidden error came in).

I think that the correct approach to this problem is to fix `Reflect.findMostSpecificInvocable` to handle the JLS 15.12.2 logic and make sure that all upstream code paths call in to this method.  The varags handling code in `Types.isSignatureVarargsAssignable` should be removed.

The fix to `Reflect.findMostSpecificInvocable` is easy enough, but really all method and constructor lookups should go through this method.  It appears that this is not consistent and some users bypass this method and thus do not handle VarArgs the right way.  I think it may be best to make some of the downstream helper methods private so that other parts of the code base do not skip the JLS 15.12.2 resolution step.

Any advice and suggestions are greatly appreciated.  Otherwise I will try to prepare a fix, additional tests and a PR for this issue.

**Comment by nickl-** [comment]
#621 RTBC

**Closed: Resolved in PR**
